pkg/agent/openflow/client.go

```diff
@@ -1189,3 +1215,69 @@ func (c *client) InstallMulticastGroup(groupID binding.GroupIDType, localReceive
 	}
 	return nil
 }
+
+// InstallMulticlusterNodeFlows installs flows to handle cross-cluster packets between a regular
+// Node and a local Gateway.
+func (c *client) InstallMulticlusterNodeFlows(gatewayID string,
+	peerConfigs map[*net.IPNet]net.IP,
+	tunnelPeerIP net.IP) error {
+	c.replayMutex.RLock()
+	defer c.replayMutex.RUnlock()
+
+	var flows []binding.Flow
+	localGatewayMAC := c.nodeConfig.GatewayConfig.MAC
+	for peerCIDR, remoteGatewayIP := range peerConfigs {
+		flows = append(flows, c.featureMulticluster.l3FwdFlowToGatewayNodeViaTun(localGatewayMAC, *peerCIDR, tunnelPeerIP, remoteGatewayIP)...)
+	}
+	return c.modifyFlows(c.featureMulticluster.cachedFlows, gatewayID, flows)
+}
+
+// InstallMulticlusterGatewayFlows installs flows to handle cross-cluster packets between Gateways.
+func (c *client) InstallMulticlusterGatewayFlows(gatewayID string,
+	peerConfigs map[*net.IPNet]net.IP,
+	tunnelPeerIP net.IP,
+	localGatewayIP net.IP,
+) error {
+	c.replayMutex.RLock()
+	defer c.replayMutex.RUnlock()
+
+	var flows []binding.Flow
+	localGatewayMAC := c.nodeConfig.GatewayConfig.MAC
+	for peerCIDR, remoteGatewayIP := range peerConfigs {
+		flows = append(flows, c.featureMulticluster.l3FwdFlowToRemoteGatewayViaTun(localGatewayMAC, *peerCIDR, tunnelPeerIP, remoteGatewayIP)...)
+		// Add SNAT flows to change cross-cluster packets' source IP to local Gateway IP.
+		flows = append(flows, c.featureMulticluster.snatConntrackFlows(*peerCIDR, localGatewayIP)...)
+	}
+	flows = append(flows, c.featureMulticluster.conntrackFlows()...)
+	return c.modifyFlows(c.featureMulticluster.cachedFlows, gatewayID, flows)
+}
+
+// InstallMulticlusterClassifierFlows adds below flows:
+// * One flow for the global virtual multicluster MAC 'aa:bb:cc:dd:ee:f0' to set its target output port as 'antrea-tun0'
+// * One flow to match MC virtual MAC 'aa:bb:cc:dd:ee:f0' in ClassifierTable table.
```

Comment:
This flow is on Gateway Node only? Please make it clear.

And I think there is a 3rd flow added (outputFlow).

Reply:
updated the comment